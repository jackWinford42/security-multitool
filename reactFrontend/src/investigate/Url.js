import React, { useState } from "react";
import { useSelector } from "react-redux";
import { Alert, Card, CardBody} from 'reactstrap';
import RamtApi from "../common/Api";
import UrlResponse from "./UrlResponse";
import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import CircularProgress from '@mui/material/CircularProgress';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogTitle from '@mui/material/DialogTitle';
import './Investigate.css';

/** Render the Url page and handle a call to the api at the investigate route
 */
export default function Url() {
  console.debug("URL COMPONENT")
  const userEmail = useSelector(st => st.currUser.email);
  const [url, setUrl] = useState("");
  const [data, setData] = useState({});
  const [loading, setLoading] = useState(false);
  const [open, setOpen] = useState(false);
  const [error, setError] = useState("");

  async function handleSubmit(evt) {
    evt.preventDefault();
    setLoading(true)
    if (!url.replace(/\s/g,"").length) setError("Blank space is not a valid url/domain.")
    else {
      setError("");
      const res = (await RamtApi.investigate({type: "url", investigate: encodeURIComponent(url), email: userEmail})).data;
      setData(res);
    }
    setLoading(false)
  }

  // Update form data to reflect change in form fields
  function handleChange(evt) {
    const { value } = evt.target;
    setUrl(value)
  }

  const handleOpen = () => {
    setOpen(true)
  };

  const handleClose = () => {
    setOpen(false)
  };

	return (
		<div className="Url">
      <Card className="formCard">
        <CardBody>
          <p>Enter an url to test its reputation.</p>
          <form className="form-inline" onSubmit={handleSubmit}>
            <label>Url:</label>
            <input
              className="form-control form-control-md"
              name="url"
              placeholder="url"
              value={url}
              onChange={handleChange}
            />
            <button type="submit" className="authButton btn btn-lg btn-primary">
              Investigate
            </button>
          </form>
        </CardBody>
      </Card>
      {loading && <CircularProgress color="secondary" />}
      {data.message && <UrlResponse data={data} open={handleOpen}/>}
      {!!error && <Alert color="danger">{error}</Alert>}
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="dialog-title"
        aria-describedby="dialog-description"
      >
        <DialogTitle id="dialog-title">
          Investigate Information
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="dialog-description">
            The safety score is a percentage based on the url/domain reputation and recent known behavior. A safety score below 25 is highly suspicious but not necessarily fraudulent.
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} autoFocus>
            close
          </Button>
        </DialogActions>
      </Dialog>
		</div>
	);
}
