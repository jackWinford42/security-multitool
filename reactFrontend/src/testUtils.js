import React from "react";
import { Provider, useDispatch } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import rootReducer from "./reducers/rootReducer";
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from "redux-thunk";

const store = createStore(
  rootReducer,
  composeWithDevTools(
    applyMiddleware(thunk)));

const demoUser = {
  email: "jackwinford@gmail.com",
  username: "captainjack",
  password: "qwerty",
  photo_url: "https://bit.ly/3C2VUgQ",
};

// const dispatch = useDispatch();
// await dispatch({type: "BEGIN_AUTH_SESSION", user: demoUser});

const UserProvider =(
    ({children}) =>
    (<Provider store={store}>
      {children}
    </Provider>)
);

export { UserProvider };