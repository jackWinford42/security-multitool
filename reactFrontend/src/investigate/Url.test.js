import React from "react";
import { render } from "@testing-library/react";
import Url from "./Url";
import { MemoryRouter } from "react-router";
import { UserProvider } from "../testUtils"

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <Url/>
      </UserProvider>
    </MemoryRouter>,
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Url/>
      </UserProvider>
    </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});