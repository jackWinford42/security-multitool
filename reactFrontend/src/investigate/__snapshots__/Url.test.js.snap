// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Url"
  >
    <div
      class="formCard card"
    >
      <div
        class="card-body"
      >
        <p>
          Enter an url to test its reputation.
        </p>
        <form
          class="form-inline"
        >
          <label>
            Url:
          </label>
          <input
            class="form-control form-control-md"
            name="url"
            placeholder="url"
            value=""
          />
          <button
            class="authButton btn btn-lg btn-primary"
            type="submit"
          >
            Investigate
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
